File: Python Fundamentals/9. Regular Expression/softuni_bar_income.py
import re

pattern = r"%(?P<name>[A-Z][a-z]+)\%[^\$\|\%\.\d]*<(?P<product>\w+)\>[^\$\|\%\.\d]*\|(?P<quantity>\d+)\|[^\$\|\%\.\d]*(?P<price>\d+\.?\d*)\$"
total_income = 0

txt = input()

while not txt == 'end of shift':
    buyer = re.match(pattern, txt)
    if buyer:
        total_price = int(buyer.group("quantity")) * float(buyer.group("price"))
        print(f"{buyer.group('name')}: {buyer.group('product')} - {total_price:.2f}")
        total_income += total_price
    txt = input()

print(f"Total income: {total_income:.2f}")

File: JavaScript Fundamentals/JavaScript Fundamentals Regex/softUniBarIncome.js
function solve(input) {
    let pattern = /%(?<name>[A-Z][a-z]+)%[^\%\.\$\|]*<(?<product>\w+)>[^\%\.\$\|]*\|(?<count>\d+)\|[^\%\.\$\|\d]*(?<price>\d+|\d+\.\d+)\$/g;
    let totalIncome = 0;

    let txt = input.shift();

    while(txt !== 'end of shift') {
        if (txt.match(pattern)) {
            let match = pattern.exec(txt);
            let {name, product, count, price} = match.groups;
            let cost = Number(count) * Number(price);
            console.log(`${name}: ${product} - ${cost.toFixed(2)}`);
            totalIncome += cost;
        }

        txt = input.shift();

    }

    console.log(`Total income: ${totalIncome.toFixed(2)}`);
}

solve(['%George%<Croissant>|2|10.3$',
'%Peter%<Gum>|1|1.3$',
'%Maria%<Cola>|1|2.4$',
'end of shift']);

solve(['%InvalidName%<Croissant>|2|10.3$',
'%Peter%<Gum>1.3$',
'%Maria%<Cola>|1|2.4',
'%Valid%<Valid>valid|10|valid20$',
'end of shift']);